src/bernsteinyang.rs

```diff
@@ -0,0 +1,351 @@
+use core::cmp::PartialEq;
+use std::ops::{Add, Mul, Sub, Neg};
+
+/// Big signed (B * L)-bit integer type, whose variables store
+/// numbers in the two's complement code as arrays of B-bit chunks.
+/// The ordering of the chunks in these arrays is little-endian.
+/// The arithmetic operations for this type are wrapping ones.
+#[derive(Clone)]
+struct ChunkInt<const B:usize, const L:usize>(pub [u64; L]);
+
+impl<const B:usize, const L:usize> ChunkInt<B,L> {
+    /// Mask, in which the B lowest bits are 1 and only they
+    pub const MASK: u64 = u64::MAX >> (64 - B);
+
+    /// Representation of 0
+    pub const ZERO: Self = Self([0; L]);
+    
+    /// Representation of 1
+    pub const ONE: Self = { 
+        let mut data = [0; L]; 
+        data[0] = 1; Self(data) 
+    };
+
+    /// Representation of -1
+    pub const MINUS_ONE: Self = Self([Self::MASK; L]);
+
+    /// Returns the result of applying B-bit right
+    /// arithmetical shift to the current number
+    pub fn shift(&self) -> Self {
+        let mut data = [0; L];
+        for i in 1..L {
+            data[i - 1] = self.0[i];
+        }
+        if self.is_negative() {
+            data[L - 1] = Self::MASK;
+        }
+        Self(data)
+    }
+
+    /// Returns the lowest B bits of the current number
+    pub fn lowest(&self) -> u64 { self.0[0] }
+
+    /// Returns "true" iff the current number is negative
+    pub fn is_negative(&self) -> bool {
+        self.0[L - 1] > (Self::MASK >> 1)
+    }
+}
+
+impl <const B:usize, const L:usize> PartialEq for ChunkInt<B,L> {
+    fn eq(&self, other: &Self) -> bool { self.0 == other.0 }
+    fn ne(&self, other: &Self) -> bool { self.0 != other.0 } 
+}
+
+impl<const B:usize, const L:usize> Add for &ChunkInt<B,L> {
+    type Output = ChunkInt<B,L>;
+    fn add(self, other: Self) -> Self::Output {
+        let (mut data, mut carry) = ([0; L], 0);
+        for i in 0..L {
+            let sum = self.0[i] + other.0[i] + carry;
+            data[i] = sum & ChunkInt::<B,L>::MASK;
```

Comment:
What if the sum is negative? You remove the negative tag there.

Reply:
The value of "sum" is never negative, since actually I operate on 2 non-negative integers in [0 .. 2 ^ (B * L) - 1] splited into B-bit non-negative chunks, but any x in [2 ^ (B * L - 1) .. 2 ^ (B * L) - 1] is considered to be the representation of -|2 ^ (B * L) - x|. Mul, Add  and Sub algorithms for the two's complement code do not care about the sign, and this is the main reason for them being used by the majority of processors. Since ChunkInt has a fixed size, it may afford to use this convenient code.